Avoid IndexOutOfRangeException in ZipArchive

peter15914 opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/109167

This pull request fixes unexpected IndexOutOfRangeException in ZipArchive.
Additional check added that array index is not out of range.
It is done in a similar way to the code in HuffmanTree.c:230:
```
if (value > 0)
{
    // prevent an IndexOutOfRangeException from array[index]
    throw new InvalidDataException(SR.InvalidHuffmanData);
}
```

Found by Linux Verification Center (linuxtesting.org).

peter15914 committed:
Avoid IndexOutOfRangeException in ZipArchive

peter15914 commented:
Hello, @ericstj , @dotnet/area-system-io-compression !
Can you review this PR?

stephentoub commented on src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/HuffmanTree.cs:
```diff
@@ -247,6 +247,12 @@ private void CreateTable()
+                            if (index >= array.Length)
+                            {
+                                // prevent an IndexOutOfRangeException from array[index]
+                                throw new InvalidDataException(SR.InvalidHuffmanData);
```
Is it possible to add a test for this?

peter15914 replied:
Hello @stephentoub, thanks for the response!
I'll try to add a test for this scenario.

stephentoub replied:
Thanks

peter15914 committed:
Merge branch 'dotnet:main' into fix-ziparchive-outofrange

akoeplinger commented:
@peter15914 FYI you'll need to wait until https://github.com/dotnet/runtime/pull/109477 is merged to be able to consume the runtime-assets changes. Once the PR is in you can merge main into your branch and you should be able to access the new file.

peter15914 committed:
Merge branch 'dotnet:main' into fix-ziparchive-outofrange

peter15914 committed:
Add Sys.IO.Compression test

Add test to cover problem with invalid zip file.
Currently an IndexOutOfRangeException is thrown when invalid zip file is opened with ZipArchive.

peter15914 commented:
@dotnet-policy-service agree

peter15914 commented:
Hello @stephentoub !
I added unit test for this scenario.
Method ZipArchive_InvalidHuffmanData is made similar to existing functions in zip_InvalidParametersAndStrangeFiles.cs.

stephentoub commented:
> Hello @stephentoub ! I added unit test for this scenario. Method ZipArchive_InvalidHuffmanData is made similar to existing functions in zip_InvalidParametersAndStrangeFiles.cs.

Thanks!

ericstj commented on src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:
```diff
@@ -871,6 +871,22 @@ public void ReadArchive_WithDiskStartNumberGreaterThanIntMax()
             Assert.Null(exception);
         }
 
+        [Theory]
+        [InlineData("HuffmanTreeException.zip")]
+        public static async Task ZipArchive_InvalidHuffmanData(string zipname)
+        {
+            string filename = bad(zipname);
```
nit, this doesn't need to be a Theory (unless you think we'll add many more test files), also add a comment
```suggestion
        /// <summary>
        /// This test checks that an InvalidDataException will be thrown when consuming a zip with bad Huffman data.
        /// </summary>
        [Fact]
        public static async Task ZipArchive_InvalidHuffmanData()
        {
            string filename = bad("HuffmanTreeException.zip");
```

ericstj reviewed (commented):
Left a small suggestion, but I'd like @carlossanlop to review too

peter15914 committed:
Update Add Sys.IO.Compression test

Change "Theory" to "Fact"

Co-authored-by: Eric StJohn <ericstj@microsoft.com>

ViktorHofer commented:
@carlossanlop PTAL

ViktorHofer committed:
Merge branch 'main' into fix-ziparchive-outofrange

carlossanlop reviewed (approved):
LGTM. The condition makes sense and the test verifies the reported scenario.

ViktorHofer commented:
/ba-g a lot of timeouts in different legs
